lib/dotcom/alerts/endpoint_stops.ex

```diff
@@ -0,0 +1,155 @@
+
+  @behaviour Behaviour
+
+  @impl Behaviour
+  def endpoint_stops(alerts, route_ids) do
+    alerts
+    |> Enum.map(&endpoint_stops_for_alert(&1, route_ids))
+    |> Enum.reject(&Kernel.is_nil/1)
+  end
+
+  defp endpoint_stops_for_alert(alert, route_ids) do
+    affected_stop_ids = alert |> Alert.get_entity(:stop)
+    affected_direction_id = alert |> affected_direction_id()
+
+    direction_names =
+      route_ids
+      |> as_actual_route_ids()
+      |> List.first()
+      |> @routes_repo.get()
+      |> Kernel.then(& &1.direction_names)
+      |> to_forward_backward_direction_map(affected_direction_id)
+
+    if MapSet.size(affected_stop_ids) < 2 do
+      nil
+    else
+      route_ids
+      |> affected_stop_lists_for(affected_direction_id, affected_stop_ids)
+      |> drop_sublists()
+      |> to_endpoints(direction_names)
+      |> case do
+        endpoints -> endpoints
+      end
+    end
+  end
+
+  defp as_actual_route_ids([]), do: []
+
+  defp as_actual_route_ids(["Green" | rest]),
+    do: GreenLine.branch_ids() ++ as_actual_route_ids(rest)
+
+  defp as_actual_route_ids([first | rest]), do: [first | as_actual_route_ids(rest)]
+
+  defp to_forward_backward_direction_map(direction_names, 0),
+    do: %{forward: direction_names[0], backward: direction_names[1]}
+
+  defp to_forward_backward_direction_map(direction_names, 1),
+    do: %{forward: direction_names[1], backward: direction_names[0]}
+
+  defp to_endpoints([], _direction_names), do: nil
+
+  defp to_endpoints([stop_list], _direction_names),
+    do: {List.first(stop_list), List.last(stop_list)}
+
+  defp to_endpoints(stop_lists, direction_names) do
+    %{
+      first_stops: stop_lists |> Enum.map(&List.first/1) |> Enum.uniq_by(& &1.id),
+      last_stops: stop_lists |> Enum.map(&List.last/1) |> Enum.uniq_by(& &1.id)
+    }
+    |> case do
+      %{first_stops: [first_stop], last_stops: [last_stop]} ->
+        {first_stop, last_stop}
+
+      %{first_stops: [first_stop], last_stops: _last_stops} ->
+        {first_stop, "#{direction_names.forward} Stops"}
+
+      %{first_stops: _first_stops, last_stops: [last_stop]} ->
+        {last_stop, "#{direction_names.backward} Stops"}
+
+      %{first_stops: _first_stops, last_stops: _last_stops} ->
+        {"#{direction_names.backward} Stops", "#{direction_names.forward} Stops"}
+    end
+
+    # dbg(first_stop_list)
+    # dbg(last_stop_list)
+  end
+
+  defp affected_stop_lists_for(route_ids, affected_direction_id, affected_stop_ids) do
+    route_ids
+    |> route_patterns_for_routes(affected_direction_id)
+    |> affected_stop_lists_per_route_pattern(affected_stop_ids)
+    |> Enum.uniq_by(&stop_ids/1)
+  end
+
+  defp affected_direction_id(alert) do
+    alert
+    |> Alert.get_entity(:direction_id)
+    |> Enum.reject(&Kernel.is_nil/1)
+    |> first_or(0)
+  end
+
+  defp first_or([first | _], _default), do: first
+  defp first_or([], default), do: default
```

Comment:
suggestion: I very recently found `List.first/2` supports a second argument that does exactly this! Use that: https://hexdocs.pm/elixir/List.html#first/2

Reply:
[Amazing](https://github.com/mbta/dotcom/pull/2558/commits/06468b37a6dce415c135d17d9df19c5ca626e981)